chain-evm/src/machine.rs

```diff
@@ -0,0 +1,248 @@
+//! # The Virtual Machine
+//!
+//! Abstractions for the EVM runtime.
+//!
+//! This module contains types that abstract away implementation details from the `evm` crate,
+//! specially those involving node and stack configurations, and runtime execution.
+//!
+//! ## Handler <- EVM Context Handler
+//! ## StackState<'config>
+//!
+
+use std::rc::Rc;
+
+use evm::{
+    backend::{Apply, ApplyBackend, Backend, Basic, Log, MemoryVicinity},
+    executor::{MemoryStackState, StackExecutor, StackSubstateMetadata},
+    Config, Context, Runtime,
+};
+use primitive_types::{H160, H256, U256};
+
+use crate::state::AccountTrie;
+
+/// Environment values for the machine backend.
+pub type Environment = MemoryVicinity;
+
+/// The context of the EVM runtime
+pub type RuntimeContext = Context;
+
+/// Top-level abstraction for the EVM with the
+/// necessary types used to get the runtime going.
+pub struct VirtualMachine {
+    environment: Environment,
+    state: AccountTrie,
+    logs: Vec<Log>,
+}
+
+impl VirtualMachine {
+    /// Creates a new `VirtualMachine` given environment values and account storage.
+    pub fn new(environment: Environment, state: AccountTrie) -> Self {
+        Self {
+            environment,
+            state,
+            logs: Default::default(),
+        }
+    }
+    /// Returns an initialized instance of `evm::executor::StackExecutor`.
+    pub fn executor<'backend, 'config>(
+        &'backend self,
+        gas_limit: u64,
+        config: &'config Config,
+    ) -> StackExecutor<'config, MemoryStackState<'backend, 'config, VirtualMachine>> {
+        let metadata = StackSubstateMetadata::new(gas_limit, config);
+        let memory_stack_state = MemoryStackState::new(metadata, self);
+        StackExecutor::new(memory_stack_state, config)
+    }
+    /// Returns an initialized instance of `evm::Runtime`.
+    pub fn runtime<'config>(
+        &self,
+        code: Rc<Vec<u8>>,
+        data: Rc<Vec<u8>>,
+        context: RuntimeContext,
+        config: &'config Config,
+    ) -> Runtime<'config> {
```

Comment:
Same as above. It would be good to have a test that exercises the abstract API to see if we can really abstract.

Reply:
Same as above.